File: auth/paseto_guide.js
const { V4 } = require('paseto');
const crypto = require('crypto');

/**
 * PASETO (Platform-Agnostic Security Tokens) for JavaScript
 * --------------------------------------------------------
 * Documentation: https://github.com/panva/paseto
 * 
 * Installation:
 *      npm install paseto
 */

// --- 1. SYMMETRIC ENCRYPTION (v4.local) ---
async function demoSymmetric() {
    console.log("--- JS Symmetric (v4.local) Demo ---");
    
    // 32-byte secret key (represented as a Buffer or Uint8Array)
    const secretKey = crypto.randomBytes(32);
    
    // Payload
    const payload = {
        sub: 'user_123',
        name: 'Nabin Hamal',
        iat: new Date().toISOString(),
    };
    
    // 1. Issue Token
    const token = await V4.encrypt(payload, secretKey);
    console.log(`Issued Token: ${token.substring(0, 40)}...`);
    
    // 2. Decrypt
    const decrypted = await V4.decrypt(token, secretKey);
    console.log(`Decrypted Payload:`, decrypted);
    console.log("\n");
}

// --- 2. ASYMMETRIC SIGNATURES (v4.public) ---
async function demoAsymmetric() {
    console.log("--- JS Asymmetric (v4.public) Demo ---");
    
    // Generate Ed25519 Key Pair
    const { privateKey, publicKey } = await V4.generateKey('public');
    
    // Payload
    const payload = {
        sub: 'admin_456',
        role: 'admin',
    };
    
    // 1. Sign
    const token = await V4.sign(payload, privateKey);
    console.log(`Signed Token: ${token.substring(0, 40)}...`);
    
    // 2. Verify
    const verified = await V4.verify(token, publicKey);
    console.log(`Verified Payload:`, verified);
    console.log("\n");
}

// Run Demos
(async () => {
    try {
        await demoSymmetric();
        await demoAsymmetric();
    } catch (err) {
        if (err.code === 'MODULE_NOT_FOUND') {
            console.error("Error: 'paseto' package not found. Please run 'npm install paseto'");
        } else {
            console.error("An error occurred:", err);
        }
    }
})();

File: auth/paseto_guide.py
import os
from datetime import datetime, timedelta, timezone
from pyseto import Pyseto, Key, Token

"""
PASETO (Platform-Agnostic Security Tokens) Guide
----------------------------------------------
PASETO is a more secure alternative to JWT. It eliminates "algorithm: none" 
vulnerabilities and provides safe defaults.

Version: v4 (Recommended)
Purpose: local (symmetric) or public (asymmetric)

Installation:
    pip install pyseto
"""

# --- 1. SYMMETRIC ENCRYPTION (v4.local) ---
# Use this when the same server issues and validates the token.
def demo_symmetric():
    print("--- Symmetric (v4.local) Demo ---")
    
    # Generate or load a 32-byte key
    # In production, store this in an environment variable
    secret_key = os.urandom(32) 
    key = Key.new(version=4, purpose="local", key=secret_key)
    
    # Payload (Claims)
    payload = {
        "sub": "user_123",
        "name": "Nabin Hamal",
        "iat": datetime.now(timezone.utc).isoformat(),
        "exp": (datetime.now(timezone.utc) + timedelta(hours=1)).isoformat(),
    }
    
    # 1. Issue Token
    token = Pyseto.encode(key, payload)
    print(f"Issued Token: {token.decode()[:40]}...")
    
    # 2. Verify and Decode
    decoded = Pyseto.decode(key, token)
    print(f"Decoded Payload: {decoded.payload}")
    print("\n")

# --- 2. ASYMMETRIC SIGNATURES (v4.public) ---
# Use this when one server issues (Private Key) and others validate (Public Key).
def demo_asymmetric():
    print("--- Asymmetric (v4.public) Demo ---")
    
    # Generate Ed25519 Key Pair
    # In production, use file-based keys (PEM/DER)
    from cryptography.hazmat.primitives.asymmetric import ed25519
    priv_key_obj = ed25519.Ed25519PrivateKey.generate()
    pub_key_obj = priv_key_obj.public_key()
    
    # Wrap in Pyseto Key objects
    signing_key = Key.from_asymmetric_key(priv_key_obj)
    verifying_key = Key.from_asymmetric_key(pub_key_obj)
    
    # Payload
    payload = {
        "sub": "admin_456",
        "role": "admin",
        "iat": datetime.now(timezone.utc).isoformat(),
    }
    
    # 1. Sign Token
    token = Pyseto.encode(signing_key, payload)
    print(f"Signed Token: {token.decode()[:40]}...")
    
    # 2. Verify Signature
    decoded = Pyseto.decode(verifying_key, token)
    print(f"Verified Payload: {decoded.payload}")
    print("\n")

if __name__ == "__main__":
    try:
        demo_symmetric()
        demo_asymmetric()
    except ImportError:
        print("Error: 'pyseto' not found. Please run 'pip install pyseto'")
    except Exception as e:
        print(f"An error occurred: {e}")
